src/render_wgpu.rs

```diff
@@ -0,0 +1,325 @@
+    inner: Option<WgpuRenderBackendInner>,
+    pipelines: Mutex<Vec<wgpu::RenderPipeline>>,
+    dpi_factor: f32,
+    frame_index: u64,
+    time: Instant,
+    time_last_frame: Instant,
+}
+
+impl WgpuRenderBackend {
+    pub fn new() -> Self {
+        Self {
+            instance: wgpu::Instance::new(wgpu::BackendBit::PRIMARY),
+            inner: None,
+            pipelines: Default::default(),
+            dpi_factor: 1.0,
+            frame_index: 0,
+            time: Instant::now(),
+            time_last_frame: Instant::now(),
+        }
+    }
+}
+
+struct WgpuRenderBackendInner {
+    surface: wgpu::Surface,
+    swapchain_desc: wgpu::SwapChainDescriptor,
+    device: wgpu::Device,
+    queue: wgpu::Queue,
+    swapchain: wgpu::SwapChain,
+    vertex_shader: wgpu::ShaderModule,
+    pipeline_layout: wgpu::PipelineLayout,
+}
+
+impl RenderBackend for WgpuRenderBackend {
+    fn init_window(&mut self, window: &dyn Any) {
+        let winit_window = window.downcast_ref::<winit::window::Window>().unwrap();
+
+        let window_size = winit_window.inner_size();
+
+        let first_init = self.inner.is_none();
+
+        let instance = &self.instance;
+
+        self.dpi_factor = winit_window.scale_factor() as f32;
+
+        let inner = self.inner.get_or_insert_with(|| {
+            let surface = unsafe { instance.create_surface(winit_window) };
+
+            let adapter =
+                pollster::block_on(instance.request_adapter(&wgpu::RequestAdapterOptions {
+                    compatible_surface: Some(&surface),
+                    ..Default::default()
+                }))
+                .ok_or_else(|| "Failed to request an adapter")
+                .unwrap();
+
+            let display_format = adapter.get_swap_chain_preferred_format(&surface);
+
+            let (device, queue) = pollster::block_on(adapter.request_device(
+                &wgpu::DeviceDescriptor {
+                    label: Some("wgpu device"),
+                    features: wgpu::Features::PUSH_CONSTANTS,
+                    limits: wgpu::Limits {
+                        max_push_constant_size: std::mem::size_of::<ShadertoyConstants>() as u32,
+                        ..Default::default()
+                    },
+                },
+                None,
+            ))
+            .unwrap();
+
+            device.on_uncaptured_error(|error| log::error!("Captured error: {}", error));
+
+            let swapchain_desc = wgpu::SwapChainDescriptor {
+                usage: wgpu::TextureUsage::RENDER_ATTACHMENT,
+                format: display_format,
+                width: window_size.width,
+                height: window_size.height,
+                present_mode: wgpu::PresentMode::Fifo,
+            };
+
+            let swapchain = device.create_swap_chain(&surface, &swapchain_desc);
+
+            let mut vertex_module_desc = wgpu::include_spirv!("shadertoy_vs.vert.spv");
+            vertex_module_desc.flags = wgpu::ShaderFlags::empty();
```

Comment:
Just FYI, we disable validation here because https://github.com/gfx-rs/naga/pull/429 isn't in wgpu 0.7.1.